# Homework2

## Monthly Retention Rate

### Anna Tigranyan

#### Libraries

In [ ]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

In [107]:
df = pd.read_csv("online_retail_II.csv")

In [108]:
df.head()

Invoice StockCode                          Description  Quantity  \
0  489434     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS        12   
1  489434    79323P                   PINK CHERRY LIGHTS        12   
2  489434    79323W                  WHITE CHERRY LIGHTS        12   
3  489434     22041         RECORD FRAME 7" SINGLE SIZE         48   
4  489434     21232       STRAWBERRY CERAMIC TRINKET BOX        24   

           InvoiceDate  Price  Customer ID         Country  
0  2009-12-01 07:45:00   6.95      13085.0  United Kingdom  
1  2009-12-01 07:45:00   6.75      13085.0  United Kingdom  
2  2009-12-01 07:45:00   6.75      13085.0  United Kingdom  
3  2009-12-01 07:45:00   2.10      13085.0  United Kingdom  
4  2009-12-01 07:45:00   1.25      13085.0  United Kingdom

<b>InvoiceDate</b>-ը դարձնենք <b>datetime</b>։

In [109]:
df["InvoiceDate"] = pd.to_datetime(df['InvoiceDate']).dt.date

Որպես <b>df1</b> մեր դատայից առանձնացնենք <b>"Customer ID"</b> և <b>"InvoiceDate"</b> սյուները,որոնք կօգտագործենք հետագա հաշվարկներում։

In [110]:
df1 = df[["Customer ID",'InvoiceDate']]
df1.head()

Customer ID InvoiceDate
0      13085.0  2009-12-01
1      13085.0  2009-12-01
2      13085.0  2009-12-01
3      13085.0  2009-12-01
4      13085.0  2009-12-01

Օգտագործելով <b>groupby</b>-ը որպես <b>start_date</b> յուրաքանչյուր Customer-ի համար վերցնենք իր այցելության առաջին օրը(կկիրառենք <b>min()</b> ֆունկցիան, որ այցելության բոլոր օրերից վերցնի առաջինը)

In [111]:
start_date = df1.groupby('Customer ID')["InvoiceDate"].min()
start_date = start_date.rename("start_day")
start_date.head()


Customer ID
12346.0    2009-12-14
12347.0    2010-10-31
12348.0    2010-09-27
12349.0    2009-12-04
12350.0    2011-02-02
Name: start_day, dtype: object

Որից հետո մեր <b>df1</b>-ին <b>merge()</b>-ի օգնությամբ ըստ <b>'Customer ID'</b>-ի <b>left join</b> իրականացնենք։

In [112]:

df1 = pd.merge(df1, start_date, left_on='Customer ID', right_index=True)
df1.head()

Customer ID InvoiceDate   start_day
0      13085.0  2009-12-01  2009-12-01
1      13085.0  2009-12-01  2009-12-01
2      13085.0  2009-12-01  2009-12-01
3      13085.0  2009-12-01  2009-12-01
4      13085.0  2009-12-01  2009-12-01

Արդյունքում յուրաքանչյուր <b>Customer ID</b>-ի համար ստանում ենք 2 սյուն՝ <b>InvoiceDate</b> և թե տվյալ հաճախորդը առաջին անգամ երբ է այցելել՝ <b>start_day</b>։

Եթե մենք այս երկու սյուների օրերը իրարից հանենք կստանանք  տվյալ հաճախորդի համար իր առաջին և հետագա այցելության միջև եղած օրերի քանակը, այսինքն՝ քանի օր անց է նորից այցելել(եթե ամսաթվերը համընկնեն կլինի 0, եթե 1 ստանքնք ուրեմն այդ հաճախորդը տվյալ start_day-ից 1 օր հետո հետ է վերադարձել )։

In [113]:
df1['days_since_start'] = (df1['InvoiceDate'] - df1['start_day']).dt.days
df1.head()

Customer ID InvoiceDate   start_day  days_since_start
0      13085.0  2009-12-01  2009-12-01                 0
1      13085.0  2009-12-01  2009-12-01                 0
2      13085.0  2009-12-01  2009-12-01                 0
3      13085.0  2009-12-01  2009-12-01                 0
4      13085.0  2009-12-01  2009-12-01                 0

Ստացանք յուրաքանչյուր հաճախորդի համար վերը նկարագրված <b>days_since_start</b> սյունը:

Այժմ որպես սկզբնական ամսաթիվ վերցնենք start_day-ի ամենափոքր ամսաթիվը։

In [114]:
date1 = df1["start_day"][0]

Վերցնենք df1 միայն այն տողերը որոնց համար <b>start_day</b>–ը հավասար է եղել <b>date1</b>-ի։

In [119]:
A1 = df1.loc[df1["start_day"] == date1]
A1.head()

Customer ID InvoiceDate   start_day  days_since_start
0      13085.0  2009-12-01  2009-12-01                 0
1      13085.0  2009-12-01  2009-12-01                 0
2      13085.0  2009-12-01  2009-12-01                 0
3      13085.0  2009-12-01  2009-12-01                 0
4      13085.0  2009-12-01  2009-12-01                 0

Հանենք <b>դուպլիկատները</b>․

In [121]:
A1 = A1.drop_duplicates()
A1.head()

Customer ID InvoiceDate   start_day  days_since_start
0           13085.0  2009-12-01  2009-12-01                 0
7739        13085.0  2009-12-03  2009-12-01                 2
74247       13085.0  2010-01-29  2009-12-01                59
394957      13085.0  2010-10-15  2009-12-01               318
619157      13085.0  2011-02-17  2009-12-01               443

Հաշվենք <b>daily retention</b>-ը։

In [117]:
x = A1.groupby('days_since_start')["start_day"].count()
x = x/x[0] * 100
x.head(10)

days_since_start
0     100.000000
1       2.564103
2       6.837607
3       2.564103
5       0.854701
6       5.128205
7       9.401709
8       5.128205
9       5.982906
10     12.820513
Name: start_day, dtype: float64

Ունեցանք, որ <b>2009-12-01</b> -ի համար <b>1</b> օր անց վերադարձել են տվյալ օրվա հաճախորդների <b>2.564103</b> տոկոսը, <b>2</b> օր անց՝ <b>6.837607</b> տոկոսը,․․․


Նման ձևով կարելի է հաշվարկել կամայական ամսաթվի համար։